### PR TITLE
DSDM Enrolment functionality and display on Patient Dashboard

### PR DESCRIPTION
**Pull request by slubwama**
https://trello.com/c/4cuG13eC/764-add-dsdm-on-art-encounter-page-and-enable-enrolling-in-a-dsdm-model-when-selected

https://trello.com/c/ogpNAWAE/762-add-a-page-to-add-dsdm-models-to-patient

https://trello.com/c/xsBM9ufU/763-change-dsdm-concept-to-coded-and-add-the-different-dsdms-program-concepts-as-answers

https://trello.com/c/8fsFbudF/761-add-dashboard-widget-to-show-programs-patient-is-enrolled-in

https://trello.com/c/GF5Uizbx/760-dsdm-abbreviate-the-dsdm-program-names

Adding widget on dashboard to show program patient is enrolled in
Adding page to enrol patient in program
Adding functionality to enrol DSDM on ART encounter card
Shortening names of DSDM programs
Changing DSDM concept to be of datatype coded

[Documentation for DSDM Concept](https://drive.google.com/open?id=122O4i43iYZQb0nRVfXoJoyltGQssU2GFJ2CLptl_nak)

**Reference to closed #357**

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/aijar/fragment/controller/PatientStabilityFragmentController.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import org.openmrs.api.PatientService;
+import org.openmrs.api.context.Context;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.fragment.FragmentModel;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.*;
```
Comment (ssmusoke) [inline omod/src/main/java/org/openmrs/module/aijar/fragment/controller/PatientStabilityFragmentController.java]: @slubwama Remove wildcard imports

File: omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.*;
+import org.openmrs.api.context.Context;
+import org.openmrs.module.aijar.fragment.controller.PatientSummaryFragmentController;
+import org.openmrs.module.htmlformentry.CustomFormSubmissionAction;
+import org.openmrs.module.htmlformentry.FormEntryContext;
+import org.openmrs.module.htmlformentry.FormEntryContext.Mode;
+import org.openmrs.module.htmlformentry.FormEntrySession;
+
+import java.util.*;
```
Comment (ssmusoke) [inline omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java]: @slubwama Remove wildcard imports

File: omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+import static org.openmrs.module.aijar.AijarConstants.GP_DSDM_CONCEPT_ID;
+import static org.openmrs.module.aijar.AijarConstants.GP_DSDM_PROGRAM_UUID_NAME;
+
+/**
+ * Enrolls patients into DSDM programs
+ */
+public class DSDSProgramSubmissionAction implements CustomFormSubmissionAction {
```
Comment (ssmusoke) [inline omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java]: @slubwama Add test cases for the program submission to help aiding review of the different scenarios

File: omod/src/main/resources/config.xml [hunk 4]
```diff
@@ -146,5 +146,16 @@
...
 
+    <globalProperty>
+        <property>ugandaemr.dsdm.programsuuid</property>
+        <description>List of Programs in DSDM</description>
+        <defaultValue>de5d54ae-c304-11e8-9ad0-529269fb1459,de5d5b34-c304-11e8-9ad0-529269fb1459,de5d5896-c304-11e8-9ad0-529269fb1459,de5d5da0-c304-11e8-9ad0-529269fb1459,de5d6034-c304-11e8-9ad0-529269fb1459</defaultValue>
+    </globalProperty>
+
+    <globalProperty>
+        <property>ugandaemr.dsdm.conceptId</property>
+        <description>A concept for DSDM</description>
```
Comment (ssmusoke) [inline omod/src/main/resources/config.xml]: @slubwama This description is not clear what is this concept used for?

File: omod/src/main/resources/config.xml [hunk 5]
```diff
@@ -146,5 +146,16 @@
         <defaultValue></defaultValue>
     </globalProperty>
 
+    <globalProperty>
+        <property>ugandaemr.dsdm.programsuuid</property>
+        <description>List of Programs in DSDM</description>
```
Comment (ssmusoke) [inline omod/src/main/resources/config.xml]: @slubwama Suggested change - comma delimited list of program uuids for DSDM (are these program or concept uuids?)

File: omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java (hunk 3 above)
Comment (slubwama) [reply omod/src/main/java/org/openmrs/module/aijar/htmlformentry/DSDSProgramSubmissionAction.java]: @ssmusoke This is Completed.

**Comment by slubwama** [comment]
@ssmusoke this PR is ready for review

**Comment by slubwama** [comment]
@ssmusoke whats the status of review??

**Comment by ssmusoke** [comment]
For new patients enrolled into care enroll them into FBIM

**Comment by ssmusoke** [comment]
1. DSDM program text has link styling that goes no where

2. ART Encounter - when a patient is unstable the community models must not be available for selection. This could be okay since there is a validation rule that stops this from happening (but better to hide the models) 
screenshot 2019-02-11 16 54 23

3. Encounter page allows the encounter date to be changed which shows today instead of the past visit date - 
screenshot 2019-02-11 17 13 19

File: omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+import java.util.Date;
+import java.util.List;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests patient enrollment into the MCH program
+ */
+public class DSDMProgramSubmissionActionTest extends BaseModuleWebContextSensitiveTest {
```
Comment (ssmusoke) [inline omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java]: @slubwama For each of the test cases, add checks for program enrollment, and program exit to verify that these dates match what is expected in the database

Comment (slubwama) [reply omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java]: @ssmusoke this is complete.

**Comment by slubwama** [comment]
@ssmusoke done all this.

**Comment by slubwama** [comment]
@ssmusoke ETA??

**Comment by ssmusoke** [comment]
1. When a patient is enrolled into care but has no ART start date then they should not be enrolled into FBIM

2. The enrollment date into FBIM should be the ART Start Date rather than the enrollment date (I think this also changes the FBIM query that currently exists) @slubwama thoughts on this? 

3. Add test cases:
- When a patient is enrolled into care with no ART start date there should be no enrollment 
- When a patient is enrolled into care with an ART start date, this should be the date of enrollment 

4. The issues in https://github.com/METS-Programme/openmrs-module-aijar/pull/365#issuecomment-462342252 have not been addressed

File: omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        when(request.getParameter("w1")).thenReturn("2017-04-01");
+        //Seleted DSDM Program in form is FTR
+        when(request.getParameter("w8")).thenReturn("165139");
+        session1.getSubmissionController().handleFormSubmission(session1, request);
+
+        session1.applyActions();
+        List<PatientProgram> patientProgramList = getPatientPrograms(patient, null,null,null,false);
+        Assert.assertEquals(service.getProgramByName("FTR"), patientProgramList.get(1).getProgram());
+        Assert.assertEquals(service.getProgramByName("FBIM"), patientProgramList.get(0).getProgram());
+        Assert.assertEquals(patientProgramList.get(1).getDateEnrolled(), patientProgramList.get(0).getDateCompleted());
```
Comment (slubwama) [inline omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java]: @ssmusoke here is one.

Comment (ssmusoke) [inline omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java]: @slubwama Break into 2 Assertions using the hardcoded date in line 72 - that way you are sure that the dates are correct as both could be null and this test still passes

Comment (slubwama) [reply omod/src/test/java/org/openmrs/module/aijar/htmlformentry/DSDMProgramSubmissionActionTest.java]: @ssmusoke that is done

**Pull request merged**